executor/new_distsql.go

```diff
@@ -286,62 +289,193 @@ type IndexLookUpExecutor struct {
+	tableHandler
 	finished chan struct{}
 }
 
+// indexHandler is used by IndexLookUpExecutor to maintain index lookup background goroutines.
+type indexHandler struct {
+	buildTableTasks func(handles []int64) []*lookupTableTask
+	wg              sync.WaitGroup
+}
+
+func (ih *indexHandler) open(kvRanges []kv.KeyRange, e *IndexLookUpExecutor, workCh chan<- *lookupTableTask, finished <-chan struct{}) error {
+	result, err := distsql.SelectDAG(e.ctx.GetClient(), e.ctx.GoCtx(), e.dagPB, kvRanges,
+		e.ctx.GetSessionVars().DistSQLScanConcurrency, e.keepOrder, e.desc, getIsolationLevel(e.ctx.GetSessionVars()), e.priority)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	result.Fetch(e.ctx.GoCtx())
+	ih.buildTableTasks = e.buildTableTasks
+	ih.wg.Add(1)
+	go func() {
+		ctx, cancel := goctx.WithCancel(e.ctx.GoCtx())
+		ih.fetchHandles(result, workCh, ctx, finished)
+		cancel()
+		if err := result.Close(); err != nil {
+			log.Error("close SelectDAG result failed:", errors.ErrorStack(err))
+		}
+		close(workCh)
+		ih.wg.Done()
+	}()
+	return nil
+}
+
+// fetchHandles fetches a batch of handles from index data and builds the index lookup tasks.
+func (ih *indexHandler) fetchHandles(result distsql.SelectResult, workCh chan<- *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
+	for {
+		handles, finish, err := extractHandlesFromIndexResult(result)
+		if err != nil {
+			workCh <- &lookupTableTask{
+				tasksErr: errors.Trace(err),
+			}
+			return
+		}
+		if finish {
+			return
+		}
+		tasks := ih.buildTableTasks(handles)
+		for _, task := range tasks {
+			select {
+			case <-ctx.Done():
+				return
+			case <-finished:
+				return
+			case workCh <- task:
+			}
+		}
+	}
+}
+
+func (ih *indexHandler) close() {
+	ih.wg.Wait()
+}
+
+// tableHandler is used by IndexLookUpExecutor to maintain table lookup background goroutines.
+type tableHandler struct {
+	taskChan    chan *lookupTableTask
+	taskCurr    *lookupTableTask
+	executeTask func(task *lookupTableTask, goCtx goctx.Context)
+}
+
+func (th *tableHandler) open(e *IndexLookUpExecutor, workCh <-chan *lookupTableTask, finished <-chan struct{}) {
+	th.executeTask = e.executeTask
+	th.taskChan = make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
+	lookupConcurrencyLimit := e.ctx.GetSessionVars().IndexLookupConcurrency
+	var wg sync.WaitGroup
+	wg.Add(lookupConcurrencyLimit)
+	for i := 0; i < lookupConcurrencyLimit; i++ {
+		ctx, cancel := goctx.WithCancel(e.ctx.GoCtx())
+		go func() {
+			th.pickAndExecTask(workCh, ctx, finished)
+			cancel()
+			wg.Done()
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(th.taskChan)
+	}()
+}
+
+func (th *tableHandler) pickAndExecTask(workCh <-chan *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
+	for {
+		select {
+		case task := <-workCh:
+			if task == nil {
+				return
+			}
+			if task.tasksErr == nil {
+				th.executeTask(task, ctx)
+				th.taskChan <- task
+			} else {
+				th.taskChan <- task
+				return
```

Comment:
Do we really need to return here?
When error occurred in `fetchHandles`, the `workCh` will be closed soon.

This can be simpler.
```
if task.tasksErr == nil {
    th.executeTask(task, ctx)
}
th.taskChan <- task
```

Reply:
If error happened, we should return.

Reply:
There are too many returns here, makes it hard to reason about.
If we remove the redundant returns, we can infer the sequence of execution easier.

Reply:
And our code style is to handle error first, this looks awkward.